lib/Sema/SemaDecl.cpp

```diff
@@ -11213,28 +11213,129 @@ void Sema::ActOnFinishKNRParamDeclarations(Scope *S, Declarator &D,
   }
 }
 
-// ActOnBoundsDecl: handle a Checked C bounds declaration for a declarator.
+static void checkBoundsDeclWithTypeAnnotation(Sema &S, DeclaratorDecl *D,
+                                             InteropTypeBoundsAnnotation *Expr,
+                                             bool IsReturnBounds) {
+  assert(D != nullptr && Expr != nullptr && !Expr->isInvalid());
+
+  QualType DeclaredTy = D->getType();
+  QualType AnnotTy = Expr->getType();
+  QualType OriginalAnnotTy = AnnotTy;
+
+  if (DeclaredTy.isNull() || AnnotTy.isNull())
+    return;
+
+  if (IsReturnBounds) {
+    assert(DeclaredTy->isFunctionType());
+    if (const FunctionType *FuncTy = DeclaredTy->getAs<FunctionType>())
+      DeclaredTy = FuncTy->getReturnType();
+    else
+      return;
+  }
+
+  if (isa<ParmVarDecl>(D))
+    AnnotTy = S.Context.getAdjustedParameterType(AnnotTy);
+
+  int DiagId = 0;
+
+  // Make a best effort to detect multiple errors at once.  This might help
+  // programmers avoid multiple rounds of fixing and recompiling programs.
+  //
+  // Track the kinds of errors that have occurred to avoid emitting errors
+  // implied by prior errors, using a bitmask.
+
+  enum ErrorType {
```

Comment:
Yes, there are other definitions of enum within function bodies in the clang code base.  An example is Sema::DiagnoseSentinelCalls in SemaExpr.cpp at line 416.